### DISCUSSION TITLE
git-lfs-transfer protocol

### DISCUSSION THREAD

**Discussion by axkibe** [git-lfs/git-lfs#5621] [General]
Hi guys with my node.js git-server implementation I finally came around to try to implement git-lfs-transfer on the SSH protocol. (would be nice to have).

For now 2 questions:

So I never implemented the locking API on https mode, and don't plan to start to use it for SSH mode either. So my welcome announcment is simply
"000eversion=1 \n0000"

On https://github.com/git-lfs/git-lfs/blob/main/docs/proposals/ssh_adapter.md

it says 

> If the server supports locking, the locking capability should be advertised, and the client may then use the lock, unlock, and list-lock commands.

I didn't so the client shouldnt come up with lock commands, but the next thing I get is
list-lock
refname=refs/heads/master

Shouldnt the client doing this if I didnt tell him I can do locking?

Anyway, I could either lie to the client and give it an empty "status 200" reply (no locks) or a "status 400" .. its not mentioned in the proposal, but I guess that should be the reply, if I don't understand a cpmmand? My instinct so far was to terminate the connection, whenever something on the protocol happens I don't understand..

Anyway, regardless what I do, the client will complain to the user:
either:

Remote "origin" does not support the Git LFS locking API. Consider disabling it with:
  $ git config lfs.https://localhost/axel.git/info/lfs.locksverify false

or in lie about it varient:

Locking support detected on remote "origin". Consider enabling it with:
  $ git config lfs.https://localhost/axel.git/info/lfs.locksverify true

So eitherway the user gets a message.. which in https never appeared, however, once I enable lfs-transfer I want my users not to notice a difference (for now I used to accept the ssh session and immediately terminate with -1 error code  as unknown command, which made the client silently switch to authenticate and https)

**Comment by axkibe** [comment]
Update, sorry, I dont think this message has much to do with git-lfs-transfer after all, now it appears on my test machine as well, I think it has to do with "localhost" as URL, I'm investigating.

**Comment by axkibe** [comment]
OK, figured it out.

Funny thing is, I do not remember implementing the locking API conscious decission from mine, I guess it may have happened after my LFS implementation, or I simply overlooked it. Anyway, all URLs in REPOSITORY_NAME/info/lfs I made no implementation for would return "501 not implemented", which dues applied to to /lock/verify .. and made the git-lfs client not complain about locking.

When I copy this logic to reply "status 501" to 'list-lock' in the git-lfs-transfer protocol, it does the same, it does not complain to the user thus exactly what I want.

Now I can go on implementing the batch and upload/download API over ssh

**Comment by bk2204** [comment]
You have a good point here that we don't read the capability and I've opened #5624 to require the capability before sending these requests.  I'll want to implement a suitable mode in Scutiger so we can test this, so I've left in in draft mode, but I expect it will require few changes.

**Comment by axkibe** [comment]
BTW: Once I replied with a "status 501" it stopped doing that. 
Which btw I guess is a reasonble reply to code for any command not understood?
